CRM/Member/BAO/MembershipType.php

```diff
@@ -91,11 +91,22 @@ public static function setIsActive($id, $is_active) {
    * @param array $ids
    *   Array contains the id (deprecated).
    *
-   *
-   * @return object
+   * @return \CRM_Member_DAO_MembershipType
+   * @throws \CiviCRM_API3_Exception
    */
   public static function add(&$params, $ids = array()) {
-    $id = CRM_Utils_Array::value('id', $params, CRM_Utils_Array::value('membershipType', $ids));
+    if (empty($params['id'])) {
+      $params['id'] = CRM_Utils_Array::value('membershipType', $ids);
+    }
+    $id = CRM_Utils_Array::value('id', $params);
```

Comment:
I think the tidiest way of expressing this part is the Phone BAO

    $hook = empty($params['id']) ? 'create' : 'edit';
    CRM_Utils_Hook::pre($hook, 'Phone', CRM_Utils_Array::value('id', $params), $params);

Reply:
Agreed, updated to match that format